File: py/mongodb/lapcounter_layout.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict, Optional

from mongodb import lapcounter_layout_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class Vec2:
    x: float = 0.0
    y: float = 0.0
    
    _proto_message: lapcounter_layout_pb2.Vec2 = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = lapcounter_layout_pb2.Vec2()

    def _populate_proto(self):
        self._proto_message.x = self.x
        self._proto_message.y = self.y

    @classmethod
    def _from_proto(cls, proto_message) -> "Vec2":
        return cls(
            x = proto_message.x,
            y = proto_message.y,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Vec2":
        message = lapcounter_layout_pb2.Vec2()
        message.ParseFromString(data)
        return cls._from_proto(message)

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Vec2":
        message = lapcounter_layout_pb2.Vec2()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class PositionAndDirection:
    position: Vec2 = None
    direction: Vec2 = None
    
    _proto_message: lapcounter_layout_pb2.PositionAndDirection = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = lapcounter_layout_pb2.PositionAndDirection()

    def _populate_proto(self):
        if self.position:
            self.position._populate_proto()
            self._proto_message.position.CopyFrom(self.position._proto_message)
        if self.direction:
            self.direction._populate_proto()
            self._proto_message.direction.CopyFrom(self.direction._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "PositionAndDirection":
        return cls(
            position = Vec2._from_proto(proto_message.position),
            direction = Vec2._from_proto(proto_message.direction),
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "PositionAndDirection":
        message = lapcounter_layout_pb2.PositionAndDirection()
        message.ParseFromString(data)
        return cls._from_proto(message)

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "PositionAndDirection":
        message = lapcounter_layout_pb2.PositionAndDirection()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class Layout:
    version: int = 0
    baseline_version: int = 0
    vehicle_id: str = ""
    device_id: str = ""
    location: str = ""
    layout: str = ""
    start_line: PositionAndDirection = None
    finish_line: PositionAndDirection = None
    sectors: List[PositionAndDirection] = field(default_factory=list)
    baseline_hash: str = ""
    start_s: float = 0.0
    sectors_s: List[float] = field(default_factory=list)
    baseline_length: float = 0.0
    
    _proto_message: lapcounter_layout_pb2.Layout = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = lapcounter_layout_pb2.Layout()

    def _populate_proto(self):
        self._proto_message.version = self.version
        self._proto_message.baseline_version = self.baseline_version
        self._proto_message.vehicle_id = self.vehicle_id
        self._proto_message.device_id = self.device_id
        self._proto_message.location = self.location
        self._proto_message.layout = self.layout
        if self.start_line:
            self.start_line._populate_proto()
            self._proto_message.start_line.CopyFrom(self.start_line._proto_message)
        if self.finish_line:
            self.finish_line._populate_proto()
            self._proto_message.finish_line.CopyFrom(self.finish_line._proto_message)
        del self._proto_message.sectors[:]
        for val in self.sectors:
            val._populate_proto()
            tmp = self._proto_message.sectors.add()
            tmp.CopyFrom(val._proto_message)
        self._proto_message.baseline_hash = self.baseline_hash
        self._proto_message.start_s = self.start_s
        del self._proto_message.sectors_s[:]
        for val in self.sectors_s:
            self._proto_message.sectors_s.append(val)
        self._proto_message.baseline_length = self.baseline_length

    @classmethod
    def _from_proto(cls, proto_message) -> "Layout":
        return cls(
            version = proto_message.version,
            baseline_version = proto_message.baseline_version,
            vehicle_id = proto_message.vehicle_id,
            device_id = proto_message.device_id,
            location = proto_message.location,
            layout = proto_message.layout,
            start_line = PositionAndDirection._from_proto(proto_message.start_line),
            finish_line = PositionAndDirection._from_proto(proto_message.finish_line),
            sectors=[PositionAndDirection._from_proto(val) for val in proto_message.sectors],
            baseline_hash = proto_message.baseline_hash,
            start_s = proto_message.start_s,
            sectors_s=[float(val) for val in proto_message.sectors_s],
            baseline_length = proto_message.baseline_length,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Layout":
        message = lapcounter_layout_pb2.Layout()
        message.ParseFromString(data)
        return cls._from_proto(message)

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Layout":
        message = lapcounter_layout_pb2.Layout()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/mongodb/lapcounter_layout.cpp
#include "mongodb/lapcounter_layout.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace MongoDb
{
Vec2::Vec2(const PbMongoDb::Vec2& protobuf) {
    x = protobuf.x();
    y = protobuf.y();
}

Vec2::operator PbMongoDb::Vec2() const {
    PbMongoDb::Vec2 ret;
    ret.set_x(x);
    ret.set_y(y);
    return ret;
}

std::string Vec2::serializeAsJsonString() const {
    PbMongoDb::Vec2 protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Vec2::serializeAsProtobufString() const {
    PbMongoDb::Vec2 protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Vec2::deserializeFromJsonString(const std::string& str) {
    PbMongoDb::Vec2 protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Vec2::deserializeFromProtobufString(const std::string& str) {
    PbMongoDb::Vec2 protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

PositionAndDirection::PositionAndDirection(const PbMongoDb::PositionAndDirection& protobuf) {
    position = protobuf.position();
    direction = protobuf.direction();
}

PositionAndDirection::operator PbMongoDb::PositionAndDirection() const {
    PbMongoDb::PositionAndDirection ret;
    *(ret.mutable_position()) = position;
    *(ret.mutable_direction()) = direction;
    return ret;
}

std::string PositionAndDirection::serializeAsJsonString() const {
    PbMongoDb::PositionAndDirection protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string PositionAndDirection::serializeAsProtobufString() const {
    PbMongoDb::PositionAndDirection protobuf(*this);
    return protobuf.SerializeAsString();
}

bool PositionAndDirection::deserializeFromJsonString(const std::string& str) {
    PbMongoDb::PositionAndDirection protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool PositionAndDirection::deserializeFromProtobufString(const std::string& str) {
    PbMongoDb::PositionAndDirection protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

Layout::Layout(const PbMongoDb::Layout& protobuf) {
    version = protobuf.version();
    baseline_version = protobuf.baseline_version();
    vehicle_id = protobuf.vehicle_id();
    device_id = protobuf.device_id();
    location = protobuf.location();
    layout = protobuf.layout();
    start_line = protobuf.start_line();
    finish_line = protobuf.finish_line();
    sectors = {protobuf.sectors().begin(), protobuf.sectors().end()};
}

Layout::operator PbMongoDb::Layout() const {
    PbMongoDb::Layout ret;
    ret.set_version(version);
    ret.set_baseline_version(baseline_version);
    ret.set_vehicle_id(vehicle_id);
    ret.set_device_id(device_id);
    ret.set_location(location);
    ret.set_layout(layout);
    *(ret.mutable_start_line()) = start_line;
    *(ret.mutable_finish_line()) = finish_line;
    *(ret.mutable_sectors()) = {sectors.begin(), sectors.end()};
    return ret;
}

std::string Layout::serializeAsJsonString() const {
    PbMongoDb::Layout protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Layout::serializeAsProtobufString() const {
    PbMongoDb::Layout protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Layout::deserializeFromJsonString(const std::string& str) {
    PbMongoDb::Layout protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Layout::deserializeFromProtobufString(const std::string& str) {
    PbMongoDb::Layout protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}